drivers/vdpa/vfe/vfe_vdpa.c

```diff
@@ -0,0 +1,1196 @@
+
+static struct vfe_vdpa_priv *
+vfe_vdpa_find_priv_resource_by_vdev(struct rte_vdpa_device *vdev)
+{
+	struct vfe_vdpa_priv *priv;
+	int found = 0;
+
+	pthread_mutex_lock(&priv_list_lock);
+	TAILQ_FOREACH(priv, &vfe_priv_list, next) {
+		if (vdev == priv->vdev) {
+			found = 1;
+			break;
+		}
+	}
+	pthread_mutex_unlock(&priv_list_lock);
+	if (!found) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s.", vdev->device->name);
+		rte_errno = EINVAL;
+		return NULL;
+	}
+	return priv;
+}
+
+static uint16_t
+vfe_get_nr_vq(struct virtio_hw *hw)
+{
+	uint16_t nr_vq = hw->max_queue_pairs * 2;
+
+	if (virtio_with_feature(hw, VIRTIO_NET_F_MQ) ||
+			virtio_with_feature(hw, VIRTIO_NET_F_RSS)) {
+		vfe_virtio_read_dev_config(hw,
+			offsetof(struct virtio_net_config, max_virtqueue_pairs),
+			&hw->max_queue_pairs,
+			sizeof(hw->max_queue_pairs));
+	} else {
+		DRV_LOG(DEBUG,
+				 "Neither VIRTIO_NET_F_MQ nor VIRTIO_NET_F_RSS are supported");
+		hw->max_queue_pairs = 1;
+	}
+
+	nr_vq = hw->max_queue_pairs * 2;
+	if (virtio_with_feature(hw, VIRTIO_NET_F_CTRL_VQ))
+		nr_vq += 1;
+
+	DRV_LOG(DEBUG,"virtio nr_vq is %d",nr_vq);
+
+	return nr_vq;
+}
+
+static int
+vfe_vdpa_get_queue_num(struct rte_vdpa_device *vdev, uint32_t *queue_num)
+{
+	struct vfe_vdpa_priv *priv =
+		vfe_vdpa_find_priv_resource_by_vdev(vdev);
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s.", vdev->device->name);
+		return -1;
+	}
+	// TO_DO:for net, should use max_virtqueue_pairs? may be is same
+	*queue_num = priv->hw_nr_virtqs;
+	DRV_LOG(DEBUG, "vid %d queue num is %d", priv->vid, *queue_num);
+	return 0;
+}
+
+static int
+vfe_vdpa_get_vdpa_features(struct rte_vdpa_device *vdev, uint64_t *features)
+{
+	struct vfe_vdpa_priv *priv =
+		vfe_vdpa_find_priv_resource_by_vdev(vdev);
+	struct virtio_hw *hw = NULL;
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s.", vdev->device->name);
+		return -1;
+	}
+	hw = &priv->vpdev.hw;
+	*features = hw->guest_features;
+	*features |= (1ULL << VHOST_USER_F_PROTOCOL_FEATURES);
+   
+	return 0;
+}
+
+static int
+vfe_vdpa_get_protocol_features(struct rte_vdpa_device *vdev,
+		uint64_t *features)
+{
+	struct vfe_vdpa_priv *priv =
+		vfe_vdpa_find_priv_resource_by_vdev(vdev);
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s.", vdev->device->name);
+		return -1;
+	}
+	*features = VFE_VDPA_PROTOCOL_FEATURES;
+	return 0;
+}
+
+static uint64_t
+hva_to_gpa(int vid, uint64_t hva)
+{
+	struct rte_vhost_memory *mem = NULL;
+	struct rte_vhost_mem_region *reg;
+	uint32_t i;
+	uint64_t gpa = 0;
+
+	if (rte_vhost_get_mem_table(vid, &mem) < 0)
+		goto exit;
+
+	for (i = 0; i < mem->nregions; i++) {
+		reg = &mem->regions[i];
+
+		if (hva >= reg->host_user_addr &&
+				hva < reg->host_user_addr + reg->size) {
+			gpa = hva - reg->host_user_addr + reg->guest_phys_addr;
+			break;
+		}
+	}
+
+exit:
+	free(mem);
+	return gpa;
+}
+
+static int
+vfe_vdpa_enable_vfio_intr(struct vfe_vdpa_priv *priv)
```

Comment:
the whole code of this function doesn't belong to vdpa driver.

it belongs to virtio driver.
Refer to the 2nd architecture diagram.
The code of this function belongs to virtio_driver module and not vdpa driver.

you need to create virtio_driver directory and put the code in that library and call those apis.

the API should be, something like

virtio_dev_interrupts_enable(struct virtio_pci_dev *dev, int nr_rings);